lsコマンドを作る4

pofkuma opened the pull request:
`-l`オプション付きのlsコマンドを作成しました。

ご確認よろしくお願いします。

pofkuma committed:
Add test for '-l' option

'-l'オプションのテストケースを追加

pofkuma committed:
Fix result when no file exists

ファイルが1つも存在しなかった時の結果を正しくする

pofkuma committed:
Make '-l' option available

'-l'オプションに対応

pofkuma commented on 05.ls/ls.rb:
```diff
@@ -14,33 +17,100 @@ def convert_layout(file_names, line_count, max_name_length)
+def format_contents_long(files, base_path)
+  total_blocks = 0
+  contents = files.map do |file|
+    query_file_properties("#{base_path}/#{file}", ->(blocks) { total_blocks += blocks })
```
ファイル属性を取得するメソッドを一つ(`query_file_properties`)にまとめたかったので、
Procを使ってブロック数の集計を行うことにしました。

summertree128 replied:
おお！Lambdaですね！lsのプラクティスでLambdaを見たのは初めてかもしれないです 💡 
面白くて良いと思います 👍

summertree128 commented on 05.ls/ls.rb:
```diff
@@ -14,33 +17,100 @@ def convert_layout(file_names, line_count, max_name_length)
+  when 'link'             then :l
+  when 'socket'           then :s
+  else raise 'unkown file type'
+  end.to_s
```
このままでもダメではないんですが、`when`とか`then`とかいっぱい書かないといけないですし、elseにどういう処理を書くかも悩ましいですし、↓のようなやり方をすることが多いと思います。
https://docs.komagata.org/5691

pofkuma replied:
ありがとうございます！
ここが使い所なんですね💡　> https://docs.komagata.org/5691
https://github.com/pofkuma/ruby-practices/pull/9/commits/bf760720ac3192b09b61c5ebfac31470208ac04e で修正いたしました。

summertree128 commented on 05.ls/ls.rb:
```diff
@@ -14,33 +17,100 @@ def convert_layout(file_names, line_count, max_name_length)
+  return "ls: #{path}: No such file or directory" unless FileTest.exist?(path)
+
+  flag = all ? File::FNM_DOTMATCH : nil
+  entries = Dir.glob(*['*', flag].compact, base: path, sort: true)
```
`*['*', flag].compact` はちょっとごちゃっとしすぎ感がありますね。

https://docs.ruby-lang.org/ja/latest/method/Dir/s/=5b=5d.html
↑によると第二引数のデフォルト値は0なので、`all`が`false`の時は`flag`に`0`を代入するようにすれば、`compact`とか`*`とかしなくて良くなって見やすくなるかなーと思いました。

pofkuma replied:
それは思いつきませんでした…ありがとうございます！
36ccd5f178579b87a87d7f8786bed7d26646ae5e で修正しました。

summertree128 commented on 05.ls/ls.rb:
```diff
@@ -14,33 +17,100 @@ def convert_layout(file_names, line_count, max_name_length)
   options = {}
-  opt.on('-a') { |boolean| options[:a] = boolean }
-  opt.on('-r') { |boolean| options[:r] = boolean }
+  opt.on('-a') { |boolean| options[:all] = boolean }
+  opt.on('-r') { |boolean| options[:reverse] = boolean }
+  opt.on('-l') { |boolean| options[:long] = boolean }
```
意味が断然わかりやすくなってナイスですね 👍

summertree128 commented on 05.ls/ls.rb:
```diff
@@ -14,33 +17,100 @@ def convert_layout(file_names, line_count, max_name_length)
+end
+
+def list_directory_contents(path, all: false, reverse: false, long: false)
+  return "ls: #{path}: No such file or directory" unless FileTest.exist?(path)
```
👍

summertree128 commented on 05.ls/ls.rb:
```diff
@@ -14,33 +17,100 @@ def convert_layout(file_names, line_count, max_name_length)
+end
+
+def query_file_properties(file, totalize)
+  filestatus = File.stat(file)
```
多分ですけど、`stat`は「統計」という意味で`status`とはまた別の単語だと思います。（ので変数名はfilestatとかが良いんじゃないかな？ということです）

pofkuma replied:
勘違いしていました、`stat`は「統計」(statistic) の略なんですね…。
`filestat`に修正しました。 > adf2bc51a4cfc0be300e3fbe91f326daa3031867

summertree128 commented on 05.ls/ls.rb:
```diff
@@ -14,33 +17,100 @@ def convert_layout(file_names, line_count, max_name_length)
+def query_file_properties(file, totalize)
+  filestatus = File.stat(file)
+  filetype = filetype_char(File.ftype(file))
+  permisions_number = permissions_string(filestatus.world_readable?)
```
`permissions_number = permissons_string`はなんだか違和感がありますね。結局numberなのかstringなのかという。

pofkuma replied:
すみません、これは直し忘れでした…😅
修正いたしました。> 1f86b5d6f79b21d31454383d08167156ef48cb62

summertree128 commented on 05.ls/ls.rb:
```diff
@@ -14,33 +17,100 @@ def convert_layout(file_names, line_count, max_name_length)
+def convert_permission_to_rwx(numbers)
+  numbers.to_i.to_s(2).rjust(3, '0').chars.map.with_index do |flag, index|
+    flag.to_i.zero? ? '-' : %i[r w x][index]
+  end.join
```
なかなか難しい処理（そして正しい処理）を短いコードで表現されていますね…！シンプルにすごいと思いました！

pofkuma replied:
ここはこだわりポイントだったので、そう言っていただけて嬉しいです…！

summertree128 commented on 05.ls/ls.rb:
```diff
@@ -14,33 +17,100 @@ def convert_layout(file_names, line_count, max_name_length)
+  end.to_s
+end
+
+def convert_permission_to_rwx(numbers)
```
多分一つの数値が渡ってくる想定だと思うので、引数は`number`という名前がいいと思いました。`numbers`だと配列とかハッシュのようなコレクションを想像してしまいます。

pofkuma replied:
仰る通りだと思いました。
9968cff で修正いたしました🙏

summertree128 commented on 05.ls/ls.rb:
```diff
@@ -14,33 +17,100 @@ def convert_layout(file_names, line_count, max_name_length)
+   [:group_name,               "#{Etc.getgrgid(filestatus.gid).name} "],
+   [:number_of_filesize,       filestatus.size],
+   [:last_modified_date,       File.ctime(file).strftime('%_2m %_2d %H:%M')],
+   [:file,                     File.basename(file)]].to_h
```
初めからハッシュとして定義しないのは何か理由があるのでしょうか？

pofkuma replied:
ハッシュで値のインデントを揃えようとしたところ、RuboCopの警告が出たので、無理矢理配列にしてしまってました…。
調べたところ、Layout/HashAlignment でスタイルが決まっている事がわかりました。
https://docs.rubocop.org/rubocop/cops_layout.html#layouthashalignment

34e49998f3cab726e795c26f224cfd9db7dcf02c のように修正してみたのですが、このようにしても問題ないでしょうか？

summertree128 replied:
@pofkuma 
なるほどですね、背景は理解できましたー。
うーん、悩ましいところですが、ここは一旦rubocop-fjordに従ってもらえますか？
組織としてコードのスタイルを定めてあるのはチームで開発していれば良くあることで、基本的にはそのスタイルに従うことになります。そこに抗ってわざわざ一部のルールを無効化するというのは普通やらないことなので、そういう癖をつけてほしくはないです。どうしても受け入れられないならコーディングスタイルの定義そのものを更改する方が健全だと思いますので。あと、言語によっては自動でフォーマットするのが当然のような文化もありますし（Goとか）、そういう時はそもそもコーディングスタイルに抗うことができません。なので、スタイルが既に存在していれば基本的にそれに従う、郷に入っては郷に従えの精神でいるのが良いと思います。

ちなみにですが、僕も以前はこのように揃える書き方を好んでしていましたね。綺麗ですし。ただやたらと長い名前のキーが追加されたりすると全部インデントし直さなければいけないデメリットもあるので一長一短という感じですね。オートフォーマットができないのも短所でしょうか。しかし見た目の美しさというのは非常に大きなメリットなので、ご自身のプロジェクトだったりスタイルの強制がないような場合ならこのような形で書いても全く問題はないと思います 💡

pofkuma replied:
承知しました。詳しく教えて下さりありがとうございます🙏
6d776cc にて修正しました。

> スタイルが既に存在していれば基本的にそれに従う、郷に入っては郷に従えの精神でいるのが良いと思います。

「リーダブルコード」にも、好みより一貫性が大切と書かれていたのを思い出しました。

なお、変更したルールでもオートフォーマットは可能でした。

pofkuma replied:
@summertree128 
承知しました。詳しく教えて下さりありがとうございます🙏
6d776cc にて修正しました。

> スタイルが既に存在していれば基本的にそれに従う、郷に入っては郷に従えの精神でいるのが良いと思います。

「リーダブルコード」にも、好みより一貫性が大切と書かれていたのを思い出しました。

 ちなみに、変更したルールでもオートフォーマットは可能でした。

summertree128 commented on 05.ls/ls.rb:
```diff
@@ -14,33 +17,100 @@ def convert_layout(file_names, line_count, max_name_length)
+    [[:number_of_links,    ->(value, width) { value.rjust(width) }],
+     [:owner_name,         ->(value, width) { value.ljust(width) }],
+     [:group_name,         ->(value, width) { value.ljust(width) }],
+     [:number_of_filesize, ->(value, width) { value.rjust(width) }]].to_h
```
これもハッシュとして定義すればいいのかと思ったのですがどうでしょうか。

pofkuma replied:
https://github.com/pofkuma/ruby-practices/pull/9#discussion_r770370247 と同様です🙏

summertree128 replied:
こちらも同様です。

summertree128 commented on 05.ls/ls.rb:
```diff
@@ -14,33 +17,100 @@ def convert_layout(file_names, line_count, max_name_length)
+
+  justfying_procs.each do |name, proc|
+    max_length = contents.map { _1.fetch(name).to_s.length }.max
+    contents.map { |content| content[name] = proc.call(content.fetch(name).to_s, max_length) }
```
引数として渡ってきた`contents`をこのように破壊的に変更してしまうと、`justify_contents_value`の呼び出し元から見たときに`contents`に何が入っているのか分かりづらくなりバグの原因に繋がります。
代わりに新しいハッシュを定義してそちらを返すようにしましょう。

pofkuma replied:
意識せず元の`contents`を変更してしまっていました。納得しました…！

301cd0372bba0072a5a730c3f8f7d3d1bb732d94 のように修正してみました。（ちょっと無理矢理かも知れません…）

summertree128 commented on 05.ls/ls_test.rb:
```diff
@@ -1,87 +1,132 @@
+  TEST_DIR = "#{__dir__}/test"
+
+  def setup
+    FileUtils.rm_r(TEST_DIR) if File.exist?(TEST_DIR)
+    FileUtils.mkdir(TEST_DIR)
```
実際にファイルを作ってテストする形に変えたんですね。
ちゃんと`teardown`で消しているので問題ないとは思いますが、`Tempfile`を使っても良さそうかなと思いました。（いや、ファイル名が変になるかも…？ちょっと自信ないです）
このままでも大丈夫ですが、余力があれば少し調べてみてください 💡 
https://docs.ruby-lang.org/ja/latest/class/Tempfile.html

pofkuma replied:
ありがとうございます！
`Tempfile`を調べて試してみましたが、ファイル名を指定する方法が見つからなかった（自動生成される）ので、このままにしたいと思います。

summertree128 reviewed (commented):
レビューしました。

相変わらずとても良いですね！今回は特に、メソッドの粒度が良い感じだなと思いました。
あと、何度でも言いますがテストがあるのが素晴らしいです ✨ 

何点かコメントしてありますのでご確認お願いしますー。

pofkuma committed:
Remove conditional branchs

条件分岐させるよりハッシュを使う方がシンプルに書けるため
https://github.com/pofkuma/ruby-practices/pull/9#discussion_r769398799

pofkuma committed:
Remove to_h

ハッシュで空白揃えをするとrubocopのチェックで警告となるため、配列で書いていた。
"EnforcedColonStyle: table"の書き方はOKのはず(bug?)

pofkuma committed:
Remove array spread

配列展開を使っていて読みづらい。
flags のデフォルト値 0 を利用する

pofkuma committed:
Fix name for stat

stat は status ではなく statistic の略と思われる

pofkuma committed:
Fix name

間違った変数名をつけていたため

pofkuma committed:
Fix name with singuler

コレクションではないので複数形ではなく単数形を使う

pofkuma committed:
Don't alter argument

引数として渡ってきた`contents`を直接変更すると他にも影響が及ぶ可能性があるため

pofkuma committed:
Comply to rubocop rules

ハッシュを"EnforcedColonStyle: table"の形式で書きたい

pofkuma committed:
Don't use fetch

fetchを使う必要性がないため

pofkuma commented:
@summertree128 
今回もレビューしてくださり、ありがとうございます‼︎
こんなに丁寧に確認していただけて、感激しております…😭✨

コメントいただいた点、それぞれ対応いたしました。
お手数おかけしますが、再レビューよろしくお願いします🙏

summertree128 commented on 05.ls/ls.rb:
```diff
@@ -14,33 +17,102 @@ def convert_layout(file_names, line_count, max_name_length)
+    owner_name:         ->(value, width) { value.ljust(width) },
+    group_name:         ->(value, width) { value.ljust(width) },
+    number_of_filesize: ->(value, width) { value.rjust(width) }
+  }.each do |name, proc|
```
流石にこの書き方はちょっと見づらいので、このハッシュは一旦変数に格納した方が良いと思います。

pofkuma replied:
ご指摘ありがとうございます。
a921448 で修正しました！

summertree128 commented on 05.ls/ls.rb:
```diff
@@ -14,33 +17,102 @@ def convert_layout(file_names, line_count, max_name_length)
+  }
+end
+
+def justify_contents_value(contents)
```
引数名の`contents`はちょっと良くないかもですね。ここだけみたら何が入ってくるのか全くわからないので、ちゃんと意味のある名前をつけた方が良いと思います。

pofkuma replied:
仰る通りだと思いました🙏
f1f1caf で修正いたしました。

summertree128 commented on 05.ls/ls.rb:
```diff
@@ -14,33 +17,102 @@ def convert_layout(file_names, line_count, max_name_length)
+end
+
+def justify_contents_value(contents)
+  justified_contents = contents.dup
+
+  {
+    number_of_links:    ->(value, width) { value.rjust(width) },
+    owner_name:         ->(value, width) { value.ljust(width) },
+    group_name:         ->(value, width) { value.ljust(width) },
+    number_of_filesize: ->(value, width) { value.rjust(width) }
+  }.each do |name, proc|
+    max_length = contents.map { _1[name].to_s.length }.max
+    justified_contents.each { |content| content[name] = proc.call(content[name].to_s, max_length) }
+  end
+  justified_contents
```
この辺り一帯ですが、構造が微妙に複雑になってしまっている感じがしなくもないですね…。

`contents`と`justified_contents`は構造が全く同じ配列ですね。これはまさしく写像(map)であり、`Array#map`メソッドを使うのに最も適した場面だと感じました。つまり、

```ruby
justified_contents = contents.map { 何かしらの処理 }
```

という形になっていれば一番綺麗だと思います。こうすれば、コードを読んでいる側からも「ああ、justified_contentsはcontentsの写像なんだな。構造は同じで中の要素が何かしら変化したんだな」というのが一眼でわかります。

…とここまで書いてから、本当にこれで綺麗に実装できるのか自信がなくなってきたので試しにちょっとコードを書いてみました。結果、あまり読みやすくならなかったのでやっぱりこのままでいいと思います。
すみません、ここまで読んでもらってなんですが、このコメントはスルーしてください 😅 

ただ、せっかくなので手元で書いてみたコードを載せておきます。contents.mapしていることだけはいいと思うんですけど、それ以外のところがなんかちょっと微妙な雰囲気（そんなにわかりやすくない）になってしまった感があります。。（まあでも頭の体操になって楽しかったです）

```ruby
  procs = {
    number_of_links:    ->(value, width) { value.rjust(width) },
    owner_name:         ->(value, width) { value.ljust(width) },
    group_name:         ->(value, width) { value.ljust(width) },
    number_of_filesize: ->(value, width) { value.rjust(width) }
  }
  procs.default = ->(value, width) { value }

  max_lengths = procs.keys.map { |k| [k, contents.map { _2.to_s.length }.max] }.to_h

  contents.map { |content|
    content.map { |key, value|
      [key, procs[key].call(value.to_s, max_lengths[key])]
    }.to_h
  }
```

ハッシュのデフォルト値（あんまり使わないけど）とか、ハッシュを`map`してキーはそのままのハッシュを作るやり方とかが参考になればなーと思いました。

pofkuma replied:
@summertree128 
詳しい解説ありがとうございます…!!
実は、私も`map`で返すやり方で書きたいと思って下記のように書いてみたのですが、イマイチな気がしてやめました…。

```ruby
  justfying_procs =
    {
      number_of_links:    ->(value, width) { value.rjust(width) },
      owner_name:         ->(value, width) { value.ljust(width) },
      group_name:         ->(value, width) { value.ljust(width) },
      number_of_filesize: ->(value, width) { value.rjust(width) }
    }

  justfying_widths =
    justfying_procs.keys.map do |key|
      max_length = contents.map { _1[key].to_s.length }.max
      [key, max_length]
    end.to_h

  contents.map do |content|
    content.map do |key, value|
      new_value = justfying_procs[key]&.call(value.to_s, justfying_widths[key]) || value
      [key, new_value]
    end.to_h
  end
```


```ruby
procs.default = ->(value, width) { value }
```
このようにハッシュのデフォルト値を作る方法は初めて知りました！

```ruby
max_lengths = procs.keys.map { |k| [k, contents.map { _2.to_s.length }.max] }.to_h
```
このように短く書けたんですね…！
コードの例を共有していただけて、とても勉強になりました🙏

summertree128 reviewed (commented):
レビューしてコメントしました！

コメントに返信する際に修正が含まれるコミットのリンクをつけるスタイルはいいですね。最近は職場で同じようにやっています。レビューする側としてはだいぶみやすいですし、これをするためにはちゃんとそれぞれの修正を別々のコミットにする必要があるので、コミット粒度を細かくすることにもつながってとても良いと思います 💡

pofkuma committed:
Comply to rubocop default rules

ハッシュの形式をデフォルト(key)に合わせる

pofkuma committed:
Use concrete name

意味をより明確にする

pofkuma committed:
Use explaining variable

一旦変数に入れた方がわかりやすいため

summertree128 reviewed (approved):
すみません、遅くなりました。 レビューOKです！
